KAFKA-12816 Added tiered storage related configs including remote log manager configs.

satishd opened the pull request:
- Added tiered storage related configs including remote log manager configs.
- Added local log retention configs to LogConfig.
- Added tests for the added configs.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
MINOR Added tiered storage related configs including remote log manager configs.
- Added tiered storage related configs including remote log manager configs.
- Added local log retention configs to LogConfig.
- Added tests for all the added configs.

satishd commented on clients/src/main/java/org/apache/kafka/common/config/TopicConfig.java:
```diff
@@ -75,6 +75,20 @@
+    public static final String REMOTE_LOG_STORAGE_ENABLE_DOC = "To enable tier storage for a topic, set `remote.storage.enable` as true. " +
+            "You can not disable this config once it is enabled. It will be provided in future versions.";
+
+    public static final String LOCAL_LOG_RETENTION_MS_CONFIG = "local.retention.ms";
```
I will update KIP-405 with `local.retention.bytes` and `local.retention.ms` properties.

satishd commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,323 @@
+     * Prefix used for properties to be passed to {@link RemoteStorageManager} implementation. Remote log subsystem collects all the properties having
+     * this prefix and passed to {@code RemoteStorageManager} using {@link RemoteStorageManager#configure(Map)}.
+     */
+    public static final String REMOTE_STORAGE_MANAGER_CONFIG_PREFIX = "remote.log.storage.manager.impl.";
```
I will update KIP about the prefixes for both `RemoteStorageManager` and `RemoteLogMetadataManager` properties.  These values are intended to be the config property, I will update the PR addressing this behacior.

satishd commented:
@junrao @kowshik Please review the PR.

satishd commented:
@junrao @kowshik Gentle reminder to review these changes.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+
+    /**
+     * Prefix used for properties to be passed to {@link RemoteStorageManager} implementation. Remote log subsystem collects all the properties having
+     * this prefix and passed to {@code RemoteStorageManager} using {@link RemoteStorageManager#configure(Map)}.
```
passed => passes

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+
+    public static final String REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP = "remote.log.manager.task.interval.ms";
+    public static final String REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_DOC = "Interval at which remote log manager runs the scheduled tasks like copy " +
+            "segments, fetch remote log indexes and clean up remote log segments.";
```
"fetch remote log indexes" should be driven by the consumer, not by the interval configuration, right?

satishd replied:
Right, nice catch.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+    public static final long DEFAULT_REMOTE_LOG_MANAGER_TASK_INTERVAL_MS = 30 * 1000L;
+
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MS_PROP = "remote.log.manager.task.retry.backoff.ms";
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MS_DOC = "It represents the wait time in milli seconds before the request is retried again.";
```
It would be useful to document that this is the initial backoff time.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MS_PROP = "remote.log.manager.task.retry.backoff.ms";
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MS_DOC = "It represents the wait time in milli seconds before the request is retried again.";
+    public static final long DEFAULT_REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MS = 30 * 1000L;
```
Why is the default the same as max?

satishd replied:
It was not intended to be the same.  It is actually 500.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MAX_MS_PROP = "remote.log.manager.task.retry.backoff.max.ms";
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MAX_MS_DOC = "The maximum amount of time in milliseconds to wait when the request " +
+            "is retried again. The retry duration will increase exponentially for each request failure up to this maximum wait interval.";
+    public static final long DEFAULT_REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MAX_MS = 30 * 1000L;
```
Why is the jitter the same as max?

satishd replied:
Jitter should be double with the default value of 0.2. It is a factor mentioned in [ExponentialBackOff](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/ExponentialBackoff.java#L30).

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+    public static final ConfigDef CONFIG_DEF = new ConfigDef();
+
+    static {
+        CONFIG_DEF.define(RemoteLogManagerConfig.REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP, BOOLEAN,
```
We could get rid of RemoteLogManagerConfig right? Ditto in other places.

satishd replied:
When moving these configs from other class, these were not optimized. Updated.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,351 @@
+            " wait time to avoid thundering herds of requests.";
+    public static final long DEFAULT_REMOTE_LOG_MANAGER_TASK_RETRY_JITTER_MS = 30 * 1000L;
+
+    public static final String REMOTE_LOG_READER_THREADS_PROP = "remote.log.reader.threads";
```
What about the copying threads? Is that separate from this?

satishd replied:
Right, it is separate from `remote.log.manager.thread.pool.size`.

junrao commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -301,6 +350,16 @@ object LogConfig {
         MessageDownConversionEnableDoc, KafkaConfig.LogMessageDownConversionEnableProp)
+
+    // RemoteLogStorageEnableProp, LocalLogRetentionMsProp, LocalLogRetentionBytesProp do not have server default
+    // config names.
```
Hmm, why is that? It's inconvenient to have to set these for each topic.

satishd replied:
In the initial version, we do not want them to enable across cluster/broker. We will enable it in a future version.

kowshik replied:
It seems useful to have the configurability to enable widely (ex: across the broker), while optionally allowing the user to only enable it at a topic level when needed.

junrao replied:
I think it's useful to add those configs at the broker level even in the first version.

satishd replied:
@junrao As we discussed offline we wanted to enable cluster level configs when the feature is complete. Filed KAFKA-12969 for tracking this change.

satishd replied:
@junrao As we discussed offline we wanted to enable cluster level configs when the feature is complete. Filed KAFKA-12969 for tracking this change.

junrao commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -827,6 +828,25 @@ class KafkaConfigTest {
+        case RemoteLogManagerConfig.REMOTE_LOG_READER_MAX_PENDING_TASKS_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
+
+//        case TopicConfig.LOCAL_LOG_RETENTION_MS_CONFIG => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -2)
+//        case TopicConfig.LOCAL_LOG_RETENTION_BYTES_CONFIG => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -2)
```
Should we uncomment these?

junrao commented on core/src/test/scala/unit/kafka/log/LogConfigTest.scala:
```diff
@@ -201,4 +206,72 @@ class LogConfigTest {
+    // Check for invalid case of localRetentionBytes > retentionBytes
+    doTestInvalidLocalLogRetentionProps(500L, 200, 100, 1000L)
+
+    // Check for invalid case of localRetentionBytes (-1 viz unlimited) > retentionBytes,
```
localRetentionMs is -1 in this case.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfigTest.java:
```diff
@@ -0,0 +1,78 @@
+        RemoteLogManagerConfig expectedRemoteLogManagerConfig
+                = new RemoteLogManagerConfig(true, "dummy.remote.storage.class", "dummy.remote.storage.class.path",
+                                             "dummy.remote.log.metadata.class", "dummy.remote.log.metadata.class.path",
+                                             "listener.name", 1024 * 1024L, 1, 60000L, 100L, 60000L, 1000L, 10, 100, rsmPrefix,
```
line too long

junrao reviewed (commented):
@satishd : Thanks for the PR. A few comments below.

satishd committed:
Addressed review comments

satishd commented:
Thanks @junrao for the review, addressed them with the latest commit.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,350 @@
+import static org.apache.kafka.common.config.ConfigDef.Type.LONG;
+import static org.apache.kafka.common.config.ConfigDef.Type.STRING;
+
+public final class RemoteLogManagerConfig {
```
Hmm, why these properties need to be defined under a separate class?
A number of other existing properties are just defined inline in `KafkaConfig`, and so could these. Example: https://github.com/apache/kafka/blob/ff3b2d67a1049c75b300547ca6ecf074b6a3ab28/core/src/main/scala/kafka/server/KafkaConfig.scala#L917-L935

satishd replied:
`KafkaConfig` become huge and it is inconvenient to go through the class too. Having a separate class with respective configs gives a better structure of the code and also readability.

kowshik commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -301,6 +350,16 @@ object LogConfig {
+    // config names.
+    logConfigDef
+      // This define method is not overridden in LogConfig as these configs do not have server defaults yet.
+      .define(RemoteLogStorageEnableProp, BOOLEAN, Defaults.RemoteLogStorageEnable, MEDIUM, RemoteLogStorageEnableDoc)
```
Don't we want this to be defined in `KafkaConfig` too so that we can enable them at broker/cluster level?

satishd replied:
I wanted to expose these configs at topic level in the initial release and we can add them at the cluster level in the later versions.

kowshik reviewed (commented):
@satishd Thanks for the PR. I had few high level questions/comments.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,350 @@
+    public static final long DEFAULT_REMOTE_LOG_MANAGER_TASK_RETRY_BACK_OFF_MAX_MS = 30 * 1000L;
+
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_JITTER_PROP = "remote.log.manager.task.retry.jitter";
+    public static final String REMOTE_LOG_MANAGER_TASK_RETRY_JITTER_DOC = "Random jitter amount applied to the `remote.log.manager.task.retry.backoff.ms` " +
```
Could we document that jitter is a relative thing?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,350 @@
+    private final long remoteLogManagerTaskIntervalMs;
+    private final long remoteLogManagerTaskRetryBackoffMs;
+    private final long remoteLogManagerTaskRetryBackoffMaxMs;
+    private final double remoteLogManagerTaskRetryJitterMs;
```
Jitter is now a relative thing, not ms. So we need to name the variable appropriately.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -0,0 +1,350 @@
+    static {
+        CONFIG_DEF.define(REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP, BOOLEAN, DEFAULT_REMOTE_LOG_STORAGE_SYSTEM_ENABLE, MEDIUM, REMOTE_LOG_STORAGE_SYSTEM_ENABLE_DOC)
+                  .define(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP, STRING, null, new ConfigDef.NonEmptyString(), MEDIUM, REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_DOC)
+                  .define(REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX_PROP, STRING, null, new ConfigDef.NonEmptyString(),
```
Could we define those new properties as internal until the feature is complete?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments.

satishd committed:
Addressed reveiw comments.
 - Added more details about `remote.log.manager.task.retry.jitter` and renames remoteLogManagerTaskRetryJitterMs to remoteLogManagerTaskRetryJitter.

satishd commented:
Thanks @junrao for the review. Addressed them with the commit 69aa4f03a266d2d95538df4bfd4eaa8b334db518.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. No more comments from me.

JDK 8 and Scala 2.12 tests seem to have exited. Do you know why?

satishd committed:
Fixed the build failure in LogConfigTest with scala 2.12
Fixed KafkaConfigTest.testFromPropsInvalid by giving the right invalid value for REMOTE_LOG_MANAGER_TASK_RETRY_JITTER_PROP.

satishd commented:
@junrao There was a build failure with Scala2.12 and a test failure. These are addressed in the latest 6e66695e87c8dd7db33cb134e83e0281e2216a62. Failures in the latest run are not related to this change.

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM

junrao reviewed (commented):
@satishd : Sorry, I forgot there was another unaddressed comment. Added below.

satishd commented:
@junrao Replied to your comment inline [here](https://github.com/apache/kafka/pull/10733#discussion_r654156492).

kowshik reviewed (approved):
@satishd Thanks for the PR. LGTM.

ccding commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -96,6 +98,45 @@ case class LogConfig(props: java.util.Map[_, _], overriddenConfigs: Set[String]
   val LeaderReplicationThrottledReplicas = getList(LogConfig.LeaderReplicationThrottledReplicasProp)
   val FollowerReplicationThrottledReplicas = getList(LogConfig.FollowerReplicationThrottledReplicasProp)
   val messageDownConversionEnable = getBoolean(LogConfig.MessageDownConversionEnableProp)
+  val remoteStorageEnable = getBoolean(LogConfig.RemoteLogStorageEnableProp)
```
Why do we use `remoteStorageEnable` but are using `RemoteLogStorageEnable` in the following place
https://github.com/satishd/kafka/blob/6e66695e87c8dd7db33cb134e83e0281e2216a62/core/src/main/scala/kafka/log/LogConfig.scala#L43

Should we rename it to `remoteLogStorageEnable`?
